src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/RestoreCommandHandler.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.ServiceBroker;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio.Internal.Contracts;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.VisualStudio.OnlineEnvironment.Client
+{
+    internal class RestoreCommandHandler
+    {
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private CancellationTokenSource _cancelBuildToken;
+
+        public RestoreCommandHandler(JoinableTaskFactory joinableTaskFactory, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+        }
+
+        public void RunSolutionRestore()
+        {
+            _joinableTaskFactory.RunAsync(async () =>
+            {
+                var cancellationTokenSource = new CancellationTokenSource();
+                _cancelBuildToken = cancellationTokenSource;
+
+                try
+                {
+                    await GetAndActivatePackageManagerOutputWindowAsync();
+
+                    await TaskScheduler.Default;
+                    IBrokeredServiceContainer serviceContainer = await _asyncServiceProvider.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+                    IServiceBroker serviceBroker = serviceContainer.GetFullAccessServiceBroker();
+
+                    var nugetSolutionService = await serviceBroker.GetProxyAsync<INuGetSolutionService>(NuGetServices.NuGetSolutionService);
+
+                    try
+                    {
+                        await nugetSolutionService.RestoreSolutionAsync(cancellationTokenSource.Token);
+                    }
+                    finally
+                    {
+                        (nugetSolutionService as IDisposable)?.Dispose();
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Only log to the activity log for now
+                    ActivityLog.LogError("NuGet Package Manager", e.Message);
+                }
+                finally
+                {
+                    cancellationTokenSource.Dispose();
+                    _cancelBuildToken = null;
```

Comment:
....another place where I'm not following 2 vars.

Reply:
See above.